shared/src/ledger/ibc/client.rs

```diff
@@ -0,0 +1,401 @@
+//! IBC validity predicate for client module
+
+use std::str::FromStr;
+
+use borsh::BorshDeserialize;
+use ibc::ics02_client::client_consensus::AnyConsensusState;
+use ibc::ics02_client::client_def::{AnyClient, ClientDef};
+use ibc::ics02_client::client_state::AnyClientState;
+use ibc::ics02_client::client_type::ClientType;
+use ibc::ics02_client::context::ClientReader;
+use ibc::ics02_client::height::Height;
+use ibc::ics24_host::identifier::ClientId;
+use ibc::ics24_host::Path;
+use tendermint_proto::Protobuf;
+
+use super::{Error, Ibc, Result, StateChange};
+use crate::ledger::storage::{self, StorageHasher};
+use crate::types::ibc::{ClientUpdateData, ClientUpgradeData};
+use crate::types::storage::{Key, KeySeg};
+
+impl<'a, DB, H> Ibc<'a, DB, H>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    pub(super) fn validate_client(
+        &self,
+        client_id: &ClientId,
+        tx_data: &[u8],
+    ) -> Result<bool> {
+        match self.get_client_state_change(client_id)? {
+            StateChange::Created => self.validate_created_client(client_id),
+            StateChange::Updated => {
+                self.validate_updated_client(client_id, tx_data)
+            }
+            _ => {
+                tracing::info!(
+                    "unexpected state change for an IBC client: {}",
+                    client_id
+                );
+                Ok(false)
+            }
+        }
+    }
+
+    /// Returns the client ID after #IBC/clients
+    pub(super) fn get_client_id(key: &Key) -> Result<ClientId> {
+        match key.segments.get(2) {
+            Some(id) => ClientId::from_str(&id.raw())
+                .map_err(|e| Error::KeyError(e.to_string())),
+            None => Err(Error::KeyError(format!(
+                "The client key doesn't have a client ID: {}",
+                key
+            ))),
+        }
+    }
+
+    fn get_client_state_change(
+        &self,
+        client_id: &ClientId,
+    ) -> Result<StateChange> {
+        let path = Path::ClientState(client_id.clone()).to_string();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a client type failed");
+        self.get_state_change(&key)
+    }
+
+    fn validate_created_client(&self, client_id: &ClientId) -> Result<bool> {
+        let client_type = match self.client_type(client_id) {
```

Comment:
It seems that there's a bit of pattern with the methods that return `Option`s. Maybe we could use the `Result` instead and put the information that is logged in the `tracing::info` into error types. Then we can use the `?` syntax so these all will one-lines. This way, the errors will also be visible in the tx result, which is better because you might not be able to see the node logs if you're submitting a tx remotely

Reply:
then we should maybe define an error type for this module instead of `use super::Error` and add `impl From<client::Error> for Error` in the `ibc/mod.rs` so we get implicit conversions

Reply:
Thank you for the suggestion!
I've modified the error handling for the client and connection.

Reply:
Thank you for applying the suggestion. If I can further add to this, what I was thinking was to return the `Result` directly from `self.client_type` and all the other methods in this scope that currently return `Option`, so then we can e.g. on this line use `let client_type = self.client_type(client_id)?;`.

Reply:
Yes, that looks nicer. However, the functions of the `ClientReader` trait are defined in `ibc-rs`. We need `*Reader` traits to reuse verification functions defined in `ibc-rs`. I'll propose changing the returned type at the mainstream.